[BEAM-3981] [WIP] Futurize and fix python 2 compatibility for coders subpackage

RobbeSneyders opened the pull request:
This pull request is the result of applying the [automatic conversion provided by the future package](http://python-future.org/overview.html#automatic-conversion-to-py2-3-compatible-code) to the coders subpackage, after which all python 2 errors were fixed. The result is python 3 styled code with python 2 compatibility.
This pull request is the first of a series in which all subpackages will be updated. We should therefore discuss the chosen approach, so we can agree on one strategy to apply throughout.

The approach I've taken, is to focus on writing python 3 code with python 2 compatibility:

- The future package provides tools to forward-port our code, while the six package focuses more on backporting. I've therefore replaced six with future everywhere it was already used.

- One of the biggest problems in porting python 2 code to python 3, is the changed handling of strings and bytes. To get a consistent behavior between versions, I have tried to rewrite everything to use the `str` and `bytes` type provided by the future.builtins package. I have not used the `from __future__ import unicode_literals` import since [its changes are too implicit and introduces a risk of subtle regressions on python 2](http://python-future.org/unicode_literals.html)

I started out with running futurize on the complete coders subpackage and then tried to fix the errors introduced by the automatic conversion. This however proved to be difficult, because it's not obvious where certain errors were introduced. 
I therefore switched to a per module approach, in which I first updated all non-test modules. This way, I could check if everything still ran with the native python 2 tests. Afterwards, I updated all test modules.

This pull request also contains updates to run_pylint.sh and tox.ini, so pylint can be run with the --py3k parameter. This should help avoid regression between the different steps of the update process.

RobbeSneyders commented on sdks/python/apache_beam/coders/stream.pyx:
```diff
@@ -39,7 +39,7 @@ cdef class OutputStream(object):
       libc.stdlib.free(self.data)
 
-  cpdef write(self, bytes b, bint nested=False):
+  cpdef write(self, b, bint nested=False):
```
The bytes type has been removed from this module for now, since a mismatch between the cython bytes type and future bytes type results in a TypeError: expected bytes, got newbytes.
I have tried replacing bytes with a memory view as explained [here](http://cython.readthedocs.io/en/latest/src/tutorial/strings.html#accepting-strings-from-python-code), but this resulted in a packaging error.
Any help on this would be appreciated.

charlesccychen replied:
<!--thread_id:cc_178620576_t; commit:57a73055b9447e7a1fe6ad674673254d9a614eeb; resolved:0-->
<!--section:context-quote-->
> **RobbeSneyders** wrote:
> The bytes type has been removed from this module for now, since a mismatch between the cython bytes type and future bytes type results in a TypeError: expected bytes, got newbytes.
> I have tried replacing bytes with a memory view as explained [here](http://cython.readthedocs.io/en/latest/src/tutorial/strings.html#accepting-strings-from-python-code), but this resulted in a packaging error.
> Any help on this would be appreciated.

<!--section:body-->
CC: @robertwb, who could provide more guidance.

aaltay commented:
cc: @tvalentyn @charlesccychen

charlesccychen commented on sdks/python/apache_beam/coders/coders_test_common.py:
```diff
@@ -17,10 +17,14 @@
 """Tests common to all coder implementations."""
 from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
```
<!--new_thread; commit:57a73055b9447e7a1fe6ad674673254d9a614eeb; resolved:0-->
Which of these is really necessary?  Do we use print in this file?

RobbeSneyders replied:
All three have been added to the top of each updated module. This is safest to avoid regression before we add full python 3 support.

charlesccychen commented on sdks/python/apache_beam/coders/coder_impl.py:
```diff
@@ -54,11 +58,6 @@
-try:
-  long        # Python 2
-except NameError:
-  long = int  # Python 3
```
<!--new_thread; commit:cc2022157865d44e8dba070a766de78853c508bc; resolved:0-->
I think we need to keep `long` for Python 2 compatibility.

RobbeSneyders replied:
The [int from the future.builtins](http://python-future.org/compatible_idioms.html#long-integers) is a subclass of python 2's long.

charlesccychen commented on sdks/python/apache_beam/coders/coders_test_common.py:
```diff
@@ -40,10 +44,11 @@
   def encode(self, x):
-    return str(x+1)
+    x = x + 1
+    return int(x).to_bytes((x.bit_length() + 7) // 8, 'big', signed=True)
```
<!--new_thread; commit:57a73055b9447e7a1fe6ad674673254d9a614eeb; resolved:0-->
What is the rationale for changing the coder behavior here to use a binary encoding instead of using a printable decimal string, for what is intended to be a simple example?  Is there something simpler we can do? (e.g. we could coerce to `str` and encode as `.encode('latin-1')`)

charlesccychen commented on sdks/python/apache_beam/coders/stream.pxd:
```diff
@@ -23,7 +23,7 @@ cdef class OutputStream(object):
   cdef size_t pos
 
-  cpdef write(self, bytes b, bint nested=*)
+  cpdef write(self, b, bint nested=*)
```
<!--new_thread; commit:57a73055b9447e7a1fe6ad674673254d9a614eeb; resolved:0-->
I'm not sure removing this type annotation is the right thing to do, for performance...

CC: @robertwb, who could provide more guidance.

RobbeSneyders replied:
This wasn't meant to be permanent. I wanted to get some feedback on this in the pr reviews and then add a commit to fix it.

charlesccychen commented on sdks/python/apache_beam/coders/typecoders_test.py:
```diff
@@ -33,14 +39,18 @@ def __init__(self, n):
   def encode(self, value):
-    return str(value.number)
+    x = value.number
+    return int(x).to_bytes((x.bit_length() + 7) // 8, 'big', signed=True)
```
<!--new_thread; commit:e3d8e1e2f5f597ff6bf0e896953f3996e74d8022; resolved:0-->
What is the rationale for changing the coder behavior here to use a binary encoding instead of using a printable decimal string, for what is intended to be a simple example?  Is there something simpler we can do? (e.g. we could coerce to `str` and encode as `.encode('latin-1')`)

RobbeSneyders replied:
That would be simpler indeed. It clashed with the cython bytes, since `str().encode('latin-1')` will return `newbytes` with the future package. 
However, seems like we need to fix this cython bytes issue anyway, so I have changed it back. This is the drawback of working on a per module basis I guess :)

charlesccychen commented on sdks/python/apache_beam/coders/proto2_coder_test_messages_pb2.py:
```diff
@@ -17,16 +17,15 @@
-import sys
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
```
<!--new_thread; commit:8f4ed891ad8e0a0b3c7da031a7618c00545b64e5; resolved:0-->
Should we revert the changes in this generated file?

Also, which of these is really necessary?  Do we use division and print in this file?

charlesccychen commented on sdks/python/apache_beam/coders/standard_coders_test.py:
```diff
@@ -74,7 +77,7 @@ class StandardCodersTest(unittest.TestCase):
               start=Timestamp(micros=(x['end'] - x['span']) * 1000),
               end=Timestamp(micros=x['end'] * 1000)),
-      'urn:beam:coders:stream:0.1': lambda x, parser: map(parser, x),
+      'urn:beam:coders:stream:0.1': lambda x, parser: list(map(parser, x)),
```
<!--new_thread; commit:5fe62966c143f8060ff75ab7edb2258f1e0df55e; resolved:0-->
Do we need this `list` call here?  This may not be intended behavior, to materialize the list, since the stream may be a long sequence that doesn't fit in memory.

RobbeSneyders replied:
I don't think we need it. Seems like I missed this one. Thanks for pointing out.

RobbeSneyders replied:
There's one test that fails without the list call. I'll look into it.

charlesccychen commented on sdks/python/apache_beam/coders/slow_coders_test.py:
```diff
@@ -16,6 +16,9 @@
 """Unit tests for uncompiled implementation of coder impls."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
```
<!--new_thread; commit:3e2ba46c6b8cf23454bb84e5c2803c05f6001e49; resolved:0-->
Which of these is really necessary?  Do we use division and print in this file?

charlesccychen commented on sdks/python/apache_beam/coders/fast_coders_test.py:
```diff
@@ -16,6 +16,9 @@
 """Unit tests for compiled implementation of coder impls."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
```
<!--new_thread; commit:3e2ba46c6b8cf23454bb84e5c2803c05f6001e49; resolved:0-->
Which of these is really necessary?  Do we use division and print in this file?

charlesccychen commented on sdks/python/apache_beam/coders/observable_test.py:
```diff
@@ -16,6 +16,9 @@
 """Tests for the Observable mixin class."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
```
<!--new_thread; commit:3e2ba46c6b8cf23454bb84e5c2803c05f6001e49; resolved:0-->
Which of these is really necessary?  Do we use division and print in this file?

charlesccychen commented on sdks/python/apache_beam/coders/coders_test.py:
```diff
@@ -14,11 +14,14 @@
-
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
```
<!--new_thread; commit:eee3edaa1aae1d97685c2465d8503c042ea46f15; resolved:0-->
Which of these is really necessary?  Do we use division and print in this file?

charlesccychen commented on sdks/python/apache_beam/coders/observable.py:
```diff
@@ -20,6 +20,11 @@
 """
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
```
<!--new_thread; commit:cffe1c29cd6256ad280677ba9c9aeb3eeffe7269; resolved:0-->
Which of these is really necessary?  Do we use division and print in this file?

charlesccychen commented on sdks/python/apache_beam/coders/coders.py:
```diff
@@ -33,6 +36,11 @@
+try:
+  import cPickle as pickle
+except ImportError:
+  import pickle
```
<!--new_thread; commit:21005b886b32241736b829e18f0aa386bbfbc13f; resolved:0-->
Please add comment that this is for Py2/3 compatibility, since in Python 3, cPickle was renamed pickle.

charlesccychen commented on sdks/python/setup.py:
```diff
@@ -69,7 +69,7 @@ def get_version():
 
 
-REQUIRED_CYTHON_VERSION = '0.26.1'
+REQUIRED_CYTHON_VERSION = '0.28.1'
```
<!--new_thread; commit:05e95b6995f0b88a53930c30459c8d81e6692219; resolved:0-->
How was this version bump determined?

charlesccychen commented on sdks/python/apache_beam/coders/coder_impl.py:
```diff
@@ -278,38 +277,38 @@ def get_estimated_size_and_observables(self, value, nested=False):
-    t = type(value)
-    if t is NoneType:
+    if value is None:
       stream.write_byte(NONE_TYPE)
```
<!--new_thread; commit:05e95b6995f0b88a53930c30459c8d81e6692219; resolved:0-->
Can you revert the change removing `t = type(value)`?  I believe this is done because `isinstance` is much slower than `is`, and this is very performance-sensitive code.

CC: @robertwb

RobbeSneyders replied:
This was changed because `is` doesn't check for subclasses. The whole future.builtins package depends on subclasses to add compatibility.
This can however be changed by depending on six again if necessary.

RobbeSneyders replied:
I have just checked the timing myself. 
`type(list()) is list` is exactly as fast as `isinstance(list(), list)`, but with an average of 5 if checks for every type call, the type check is 2.5x faster.

charlesccychen commented on sdks/python/apache_beam/coders/coder_impl.py:
```diff
@@ -199,7 +198,7 @@ def __init__(self, coder, step_label):
 
   def _check_safe(self, value):
-    if isinstance(value, (str, six.text_type, long, int, float)):
+    if isinstance(value, (str, bytes, int, float)):
```
<!--new_thread; commit:05e95b6995f0b88a53930c30459c8d81e6692219; resolved:0-->
I think we need to keep `long` for Python 2 compatibility.

charlesccychen commented on sdks/python/apache_beam/coders/coder_impl.py:
```diff
@@ -27,10 +27,14 @@
-
-import six
+from builtins import chr
+from builtins import int
```
<!--new_thread; commit:05e95b6995f0b88a53930c30459c8d81e6692219; resolved:0-->
For my edification, is this the recommended way of using these types / methods in python 3?  Are they no longer in the global namespace?  It seems a bit verbose to have to import `object` like this in every file.

RobbeSneyders replied:
This is for compatibility. On python 2, they import from future.builtins, on python 3 it has no effect.

RobbeSneyders commented:
Thanks for the review @charlesccychen.

Some general points based on your feedback and my answers:

- The imports:
`from __future__ import absolute_import`
`from __future__ import division`
`from __future__ import print_function`
were added at the top of each updated module to prevent regression before full python 3 support is added. This way no new code can be added using for instance the old python 2 division. Another benefit is the consistency of division and print across modules.

- `from builtins import ...` imports from future.builtins on python 2 and has no effect on python 3. future.builtins contains a bunch of backported python 3 builtins for compatibility.
I actually think it's best to add `from builtins import *` to every module. Just like the \_\_future\_\_ imports, this helps avoid regression and adds consistency. This however still breaks some tests. I'll try to fix these and update the pull request later.

- The bytes type annotation was removed in the stream cython files because it was breaking due to a mismatch between the cython bytes type and the future bytes type. However, this is not meant to be merged this way, but I wanted to submit the pull request with working code to get some feedback on this. I have tried replacing bytes with a memory view as explained [here](http://cython.readthedocs.io/en/latest/src/tutorial/strings.html#accepting-strings-from-python-code), but this resulted in a packaging error. Any help on this is appreciated.

- The Cython version was upgraded from 0.26.1 to 0.28.1 because of an incompatibility between cython and future types. I have not noticed any backward incompatibility.

- The is type checks were replaced by isinstance checks because the future.builtins are all subclasses of the standard python classes. However, works a lot slower. I could revert this change if I use six again for compatibility. The drawback is that `str` and `bytes` will work different across modules.

asfgit commented:
FAILURE
 
--none--

RobbeSneyders commented:
I've added some changes. Most notable:

- Replace bytes with memoryview in stream cython files. This also works with subtypes of bytes like the future.builtins bytes type.

- Revert isinstance checks to typechecks for performance with the use of past.builtins.

I've also added the applied approach to the [Python 3 proposal document](https://docs.google.com/document/d/1xDG0MWVlDKDPu_IW9gtMvxi2S9I0GB0VDTkPhjXT0nE/edit). It would be great to get some feedback on this so we can start moving forward with the other subpackages.

@aaltay @charlesccychen

RobbeSneyders commented:
I ran into some problems while applying the same strategy to some more subpackages (mostly the typehints package). Most of these problems are caused by the use of future.builtins str, bytes and int. 

I've therefore decided to change the used approach, which can be found in the new pull request #5053. A new pull request was made because a lot of the comments in this thread are now irrelevant.

The new approach is unfortunately a bit less resistant to regression, but a better fit for the beam project and I suspect we will be able to move forward more quickly.

Closing this PR in favor of #5053
